src/relay/ir/module.cc

```diff
@@ -137,5 +158,88 @@ TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
       p->stream << "ModuleNode( " << node->functions << ")";
     });
 
+void ModuleNode::ReportError(const Error & err) {
+  this->err_reporter.Report(err);
+}
+
+void ModuleNode::ReportError(const GlobalVar& global, const Expr& expr, const Error& err) {
+  this->err_reporter.ReportAt(global, expr, err);
+}
+
+template<typename T, typename U>
+using NodeMap = std::unordered_map<T, U, NodeHash, NodeEqual>;
+
+[[noreturn]] void ModuleNode::RenderErrors(bool use_color) {
+  // First we pick an error reporting strategy for each error.
+  // TODO(@jroesch): Spanned errors are currently not supported.
+  for (auto err : this->err_reporter.errors_) {
+    CHECK(!err.sp.defined()) << "attempting to use spanned errors, currently not supported";
+  }
+
+  NodeMap<GlobalVar, NodeMap<NodeRef, std::string>> error_maps;
+
+  // Set control mode in order to produce colors;
+  if (use_color) {
+    rang::setControlMode(rang::control::Force);
+  }
+
+  for (auto pair : this->err_reporter.node_to_gv_) {
+    auto node = pair.first;
+    auto global = Downcast<GlobalVar>(pair.second);
+
+    auto has_errs = this->err_reporter.node_to_error_.find(node);
+
+    CHECK(has_errs != this->err_reporter.node_to_error_.end());
+
+    auto error_indicies = has_errs->second;
+
+    std::stringstream err_msg;
+
+    for (auto index : error_indicies) {
+      err_msg << rang::fg::red;
+      err_msg << this->err_reporter.errors_[index].what();
+      err_msg << rang::fg::reset;
+    }
+
+    // Setup error map.
+    auto it = error_maps.find(global);
+    if (it != error_maps.end()) {
+      it->second.insert({ node, err_msg.str() });
+    } else {
+      error_maps.insert({ global, { { node, err_msg.str() }}});
+    }
+  }
+
+  std::stringstream annotated_prog;
+  annotated_prog << rang::style::bold <<
+  "\nError(s) occured, the below program has " \
+  "been annotated with them:\n" << rang::style::reset;
+
+  for (auto pair : error_maps) {
+    auto global = pair.first;
+    auto err_map = pair.second;
+    auto func = this->Lookup(global);
+
+    annotated_prog << RelayPrint(func, false, [=](tvm::relay::Expr expr) {
+      auto it = err_map.find(expr);
+      if (it != err_map.end()) {
+        return it->second;
+      } else {
+        return std::string("");
+      }
+    });
+  }
+
+  auto msg = annotated_prog.str();
+
+  if (use_color) {
+    rang::setControlMode(rang::control::Auto);
+  }
+
+  LOG(FATAL) << annotated_prog.str() << std::endl;
```

Comment:
@tqchen Should we display the stack trace here using `LOG(FATAL)` or should we use `std::cout`? I feel like the stack trace makes things more noisy.

Reply:
Stacktrace usually provides additional information when error message is incomplete, I would suggest we keep it for now, eventually, after we think that the error message is robust enough, we can change it to throw